File: bridge/PipeUpstreamTransport.ts
import type { CdpCommandMessage } from "../types/modcdp.js";
import { UpstreamTransport, type UpstreamTransportConfig } from "./UpstreamTransport.js";

export class PipeUpstreamTransport extends UpstreamTransport {
  readonly mode = "pipe" as const;
  readonly endpoint_kind = "raw_cdp" as const;
  declare url: string;
  private buffer = "";
  private connected = false;

  private pipe_read: NodeJS.ReadableStream | null;
  private pipe_write: NodeJS.WritableStream | null;

  constructor(
    pipe_read: NodeJS.ReadableStream | null = null,
    pipe_write: NodeJS.WritableStream | null = null,
    url = "pipe://unknown",
  ) {
    super();
    this.pipe_read = pipe_read;
    this.pipe_write = pipe_write;
    this.url = url;
  }

  update(config: UpstreamTransportConfig = {}) {
    this.pipe_read = config.pipe_read ?? this.pipe_read;
    this.pipe_write = config.pipe_write ?? this.pipe_write;
    this.url = config.cdp_url ?? config.url ?? this.url;
    return this;
  }

  getLauncherConfig() {
    return { remote_debugging: "pipe" as const };
  }

  async connect() {
    if (!this.pipe_read || !this.pipe_write) {
      throw new Error("upstream.mode=pipe requires launcher-provided remote-debugging pipe handles.");
    }
    if (this.connected) return;
    this.connected = true;
    this.pipe_read.on("data", (chunk) => this.read(chunk));
    this.pipe_read.on("end", () => this.emitClose(new Error("CDP pipe closed")));
    this.pipe_read.on("error", () => this.emitClose(new Error("CDP pipe error")));
    this.pipe_write.on("error", () => this.emitClose(new Error("CDP pipe write error")));
  }

  send(message: CdpCommandMessage) {
    if (!this.pipe_write || !this.connected) throw new Error("CDP pipe is not connected.");
    this.pipe_write.write(`${JSON.stringify(message)}\0`);
  }

  async close() {
    try {
      this.pipe_write?.end();
    } catch {}
    try {
      (this.pipe_read as { destroy?: () => void } | null)?.destroy?.();
    } catch {}
    this.connected = false;
  }

  private read(chunk: Buffer | string) {
    this.buffer += Buffer.isBuffer(chunk) ? chunk.toString("utf8") : chunk;
    for (;;) {
      const end = this.buffer.indexOf("\0");
      if (end < 0) return;
      const message = this.buffer.slice(0, end);
      this.buffer = this.buffer.slice(end + 1);
      if (message) this.parseAndEmitRecv(message);
    }
  }
}

File: client/python/modcdp/PipeUpstreamTransport.py
from __future__ import annotations

import json
import threading
from typing import Any

from .UpstreamTransport import UpstreamTransport


class PipeUpstreamTransport(UpstreamTransport):
    mode = "pipe"
    endpoint_kind = "raw_cdp"

    def __init__(self, pipe_read: Any | None = None, pipe_write: Any | None = None, url: str = "pipe://unknown") -> None:
        super().__init__()
        self.pipe_read = pipe_read
        self.pipe_write = pipe_write
        self.url = url
        self._thread: threading.Thread | None = None
        self._connected = False
        self._closed = False

    def update(self, config: dict[str, Any] | None = None) -> "PipeUpstreamTransport":
        config = config or {}
        self.pipe_read = config.get("pipe_read") or self.pipe_read
        self.pipe_write = config.get("pipe_write") or self.pipe_write
        self.url = str(config.get("cdp_url") or config.get("url") or self.url)
        return self

    def getLauncherConfig(self) -> dict[str, Any]:
        return {"remote_debugging": "pipe"}

    def connect(self) -> None:
        if self.pipe_read is None or self.pipe_write is None:
            raise RuntimeError("upstream.mode='pipe' requires launcher-provided remote-debugging pipe handles.")
        if self._connected:
            return
        self._connected = True
        self._closed = False
        self._thread = threading.Thread(target=self._read_loop, daemon=True)
        self._thread.start()

    def send(self, message: dict[str, Any]) -> None:
        if not self._connected or self.pipe_write is None:
            raise RuntimeError("CDP pipe is not connected.")
        self.pipe_write.write(json.dumps(message).encode() + b"\0")
        self.pipe_write.flush()

    def close(self) -> None:
        self._closed = True
        for pipe in (self.pipe_read, self.pipe_write):
            try:
                if pipe is not None:
                    pipe.close()
            except Exception:
                pass
        self._connected = False

    def _read_loop(self) -> None:
        buffer = b""
        try:
            while not self._closed and self.pipe_read is not None:
                chunk = self.pipe_read.read(1)
                if not chunk:
                    break
                buffer += chunk
                if b"\0" not in buffer:
                    continue
                raw, buffer = buffer.split(b"\0", 1)
                if raw:
                    self._parse_and_emit_recv(raw)
        except Exception as error:
            if not self._closed:
                self._emit_close(error if isinstance(error, Exception) else Exception(str(error)))
